core/WASM_core.md

```diff
@@ -0,0 +1,352 @@
+# Summary
+
+This proposal intends to outline the shape of the interfaces and implementation work necessary
+to make Core WASM-compilable.
+
+# Motivation
+
+A few problems a WASM Core Can solve, in no particular order:
+
+* Allow Go and Java to be ported to run on top of Core. *Potentially* without the downsides of CGo
+  and JNI. More on that later.
+* No longer any need to ship compiled Core binaries of various architectures with our existing SDKs.
+  This is both a nice plus for us, in terms of reducing CI complexity, but also a substantial size
+  reduction in our packages.
+* Workers in the browser
+* Hosted workers (also possible without, but with, there are some nice benefits)
+* A simpler way to manage callbacks into the lang layer, enabling things like user-defined metrics
+  exporters, lower-level interceptors, etc. This is certainly doable without WASM, but is probably
+  a bit easier (and, necessary) with it.
+* Auto-updates to core logic. This isn't detailed in this doc, but is a potentially very compelling
+  feature. Imagine a worker that can update core logic without needing to be restarted.
+* Allow the use of lang-native gRPC clients (though core can still provide things like retry and 
+  backoff logic, etc.)
+
+
+# Detailed design
+
+The first and largest chunk of work involved in implementation is more or less pure grunt work. 
+I need to switch Core to compiling with the wasm32-unknown-unknown target, and then fix all the
+compiler errors.
+
+There are a few predictable problems we'll definitely have to deal with:
+
+The largest hiccup is the use of network (gRPC) calls. There is more than one way we could skin this
+cat. Luckily (intentionally), all of this is contained to the `temporal-client` crate.
+
+The next major source of side effects come from the tracing and metrics exporting facilities for
+OTel and Prometheus.
+
+Additionally, there is the code dealing with downloading and starting the CLI / Java test server.
```

Comment:
I think this can be removed from the WASM version of core. If it needs to be moved out of crate or put behind a "feature" so be it. The only main benefit of sharing this code is for the HTTP and disk and process features which don't apply to WASM core. Only langs using Rust get benefit from this, other langs (e.g. Go) have their own download/extract/run code.

Reply:
Well the main point, the same point of all of core, is to just have fewer places we're duping logic. Where to download from, already downloaded checking, etc etc. In the long run, it's definitely worth centralizing to the extent we can actually save code by doing so. It's also not much of a priority.

Reply:
But some things we remove duplication of "logic" that is WASM-able and some we remove duplication of "logic" that is just system stuff. This download, extract, run thing is good for those using Rust, not good for anyone else like WASM users.